cmds/init/init.go

```diff
@@ -169,46 +164,88 @@ func main() {
+	if present && boolErr == nil && systemdEnabled == true {
+		v := cmdList[0]
+		debug("Exec %v", v)
+		if err := syscall.Exec(v, []string{v}, envs); err != nil {
+			log.Printf("Lucky you, systemd failed: %v", err)
+		}
+		// well, what a shame.
+		cmdList = cmdList[1:]
+		cmdCount++
+	}
+	// If the systemd failed for some reason, we might as well do the other things.
+
 	for _, v := range cmdList {
-		if _, err := os.Stat(v); !os.IsNotExist(err) {
-
-			noCmdFound = false
-
-			initFlags := cmdline.GetInitFlagMap()
-			// systemd gets upset when it discovers it isn't really process 1
-			systemd, present := initFlags["systemd"]
-			systemdEnabled, boolErr := strconv.ParseBool(systemd)
-			if !present || boolErr != nil || systemdEnabled == false {
-				cmd = exec.Command(v)
-				cmd.Env = envs
-				cmd.Stdin = os.Stdin
-				cmd.Stderr = os.Stderr
-				cmd.Stdout = os.Stdout
-				if *test {
-					cmd.SysProcAttr = &syscall.SysProcAttr{Cloneflags: cloneFlags}
-				} else {
-					cmd.SysProcAttr = &syscall.SysProcAttr{Setctty: true, Setsid: true, Cloneflags: cloneFlags}
-				}
-				debug("Run %v", cmd)
-				if err := cmd.Run(); err != nil {
-					log.Print(err)
-				}
+		if _, err := os.Stat(v); os.IsNotExist(err) {
+			continue
+		}
+
+		// inito is (optionally) created by the u-root command when
+		// the u-root initramfs is merged with an existing initramfs that has
+		// a /init. The name inito means "original /init"
+		// There may be an inito if we are building on
+		// an existing initramfs. All initos need their
+		// own pid space.
+		var cloneFlags uintptr
+		if v == "/inito" {
+			cloneFlags = uintptr(syscall.CLONE_NEWPID)
+		}
+
+		cmdCount++
+		cmd = exec.Command(v)
+		cmd.Env = envs
+		cmd.Stdin, cmd.Stdout, cmd.Stderr = os.Stdin, os.Stdout, os.Stderr
+		if *test {
+			cmd.SysProcAttr = &syscall.SysProcAttr{Cloneflags: cloneFlags}
+		} else {
+			cmd.SysProcAttr = &syscall.SysProcAttr{Setctty: true, Setsid: true, Cloneflags: cloneFlags}
+		}
+		debug("Run %v", cmd)
+		if err := cmd.Start(); err != nil {
+			log.Printf("Error starting %v: %v", v, err)
+			continue
+		}
+		for {
+			var s syscall.WaitStatus
+			var r syscall.Rusage
+			if p, err := syscall.Wait4(-1, &s, 0, &r); p == cmd.Process.Pid {
+				debug("Shell exited, exit status %d", s.ExitStatus())
+				break
+			} else if p != -1 {
+				debug("Reaped PID %d, exit status %d", p, s.ExitStatus())
 			} else {
-				debug("Exec %v", v)
-				if err := syscall.Exec(v, []string{v}, envs); err != nil {
-					log.Print(err)
-				}
+				debug("Error from Wait4 for orphaned child: %v", err)
+				break
 			}
 		}
-		// only the first init needs its own PID space.
-		cloneFlags = 0
+		if err := cmd.Process.Release(); err != nil {
+			log.Printf("Error releasing %v:%v", v, err)
+		}
 	}
-
-	if noCmdFound {
+	if cmdCount == 0 {
 		log.Printf("init: No suitable executable found in %+v", cmdList)
 	}
 
+	// We need to reap all children before exiting.
+	go func() {
```

Comment:
Since there is no synchronization, this goroutine is not guaranteed to reap all the leftover children. The main function could exit first which terminates all goroutines. The reaping which happens above should be enough.

I think it would be fine not to have this goroutine because the kernel panics when PID 1 exits (Linux, not sure about other kernels). When PID 1 exits in a namespace, Linux will SIGKILL all the processes in the namespace. If init is not run as PID 1, there won't be any orphans to reap anyways. In all cases, there is no need to reap the children when init will end imminently.

Reply:
I'm still not seeing it. If a child of init kicks off a subprocess, which outlasts the child, we'll blow out of the loop above with children still active. 

Don't we want to wait for those other children?

Reply:
I see what you mean. You want this to wait for all the children -- orphaned or not -- to exit before continuing.

In that case, I think this should be synchronous, so the for loop shouldn't be a separate goroutine. Keep calling `wait4` until `-1` is returned indicating there are no children left. Then continue to line 249.

Also, it might be useful to place a log message before this saying "waiting for orphaned children..."

Reply:
ah right, thanks!

Reply:
done